### PR TITLE
PLANNER-1953 Cover machine reassignment CSP by tests

### PR DESCRIPTION
**Pull request by rsynek**
Does NOT fully fix the CS provider - see the constraints throwing `UnsupportedOperationException`.

### CODE REVIEW

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 1]
```diff
@@ -145,74 +158,121 @@ private Constraint transientUsage(ConstraintFactory factory) {
...
+         * sumLong((triple, pair, processAssignment) -> pair.getRight().getMaximumCapacity() - getUsage(processAssignment,
+         * pair.getRight().getResource()))
+         * )
+         * .filter((originTriple, pair, targetAvailability) -> originTriple.right > 0 && targetAvailability <
+         * pair.left.getMultiplicand() * originTriple.right)
+         * .penalizeLong(MrConstraints.BALANCE_COST, HardSoftLongScore.ONE_SOFT, (originTriple, pair, targetAvailability) -> {
+         * System.out.println(originTriple + ", " + pair + ", " + targetAvailability);
+         * return (targetAvailability - pair.left.getMultiplicand() * originTriple.right) * pair.left.getWeight();
+         * });
+         */
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java:229]: Let's not merge this monstrosity. I have it stored for my own use.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 2]
```diff
@@ -145,74 +158,121 @@ private Constraint transientUsage(ConstraintFactory factory) {
...
-    private long balanceCost(MrBalancePenalty penalty, MrMachine machine, long originalUsage, long targetUsage) {
-        long originalAvailability =
-                machine.getMachineCapacity(penalty.getOriginResource()).getMaximumCapacity() - originalUsage;
-        long targetAvailability =
-                machine.getMachineCapacity(penalty.getTargetResource()).getMaximumCapacity() - targetUsage;
-        long balanceCost = penalty.getMultiplicand() * originalAvailability - targetAvailability;
-        return Math.max(0, balanceCost);
+    private long getUsage(MrProcessAssignment processAssignment, MrResource resource) {
+        // System.out.println(String.format("Usage of resource %d by a process %s is: %d",
+        //         resource.getIndex(), processAssignment.getProcess(), processAssignment.getUsage(resource)));
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java:234]: Again, let's remove this.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 3]
```diff
@@ -145,74 +158,121 @@ private Constraint transientUsage(ConstraintFactory factory) {
...
-        return factory.from(MrBalancePenalty.class)
-                .join(MrProcessAssignment.class)
-                .groupBy((penalty, processAssignment) -> penalty, (penalty, processAssignment) -> processAssignment.getMachine(),
-                        sumLong((penalty, processAssignment) -> processAssignment.getUsage(penalty.getOriginResource())),
-                        sumLong((penalty, processAssignment) -> processAssignment.getUsage(penalty.getTargetResource()))
-                )
-                .penalizeLong(MrConstraints.BALANCE_COST, HardSoftLongScore.ONE_SOFT, this::balanceCost);
+    protected Constraint balanceCost(ConstraintFactory factory) {
+        throw new UnsupportedOperationException("Not yet implemented.");
+        /*
```
Comment (rsynek) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java:181]: "Interesting" behavior of the eclipse formatter: the block comment has leading asterisk on each line. As a result, IDE cannot automatically uncomment such a block.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 4]
```diff
@@ -61,58 +63,72 @@
...
+    /*
+     * More elegant solution, which is not compatible with Drools due to uninitialized entities:
+     *
+     * protected Constraint serviceConflict(ConstraintFactory factory) {
+     * return factory.fromUniquePair(MrProcessAssignment.class,
+     * equal(MrProcessAssignment::getMachine),
+     * equal(MrProcessAssignment::getService)
+     * ).penalize(MrConstraints.SERVICE_CONFLICT, HardSoftLongScore.ONE_HARD);
+     *
      */
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java:89]: I'd rather not confuse readers with this.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 5]
```diff
@@ -61,58 +63,72 @@
...
+         * equal((service, location) -> service, MrProcessAssignment::getService),
+         * equal((service, location) -> location, MrProcessAssignment::getLocation))
+         * .groupBy((service, location) -> service,
+         * ConstraintCollectors.countDistinct((service, location) -> location)
+         * )
+         * .filter((service, distinctLocationCount) -> distinctLocationCount < service.getLocationSpread())
+         * .penalizeLong(MrConstraints.SERVICE_LOCATION_SPREAD, HardSoftLongScore.ONE_HARD,
+         * (service, distinctLocationCount) -> service.getLocationSpread() - distinctLocationCount);
+         *
+         */
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java:116]: Dtto.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java (hunk 3 above)
Comment (triceo) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java:181]: Maybe that's fixable?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 6]
```diff
@@ -61,58 +62,48 @@
...
      */
-    private Constraint serviceLocationSpread(ConstraintFactory factory) {
-        return factory.from(MrProcessAssignment.class)
-                .groupBy(MrProcessAssignment::getService,
-                        ConstraintCollectors.countDistinct(MrProcessAssignment::getLocation))
-                .filter((service, distinctLocationCount) -> service.getLocationSpread() > distinctLocationCount)
-                .penalize(MrConstraints.SERVICE_LOCATION_SPREAD,
-                        HardSoftLongScore.ONE_HARD);
+    protected Constraint serviceLocationSpread(ConstraintFactory factory) {
+        throw new UnsupportedOperationException("ConstraintCollectors.countDistinct should not count null as a distinct value");
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java:91]: Which MrProcessAssignment have a null location? The `from` already filters out all those that doesn't have a machine. Are there machines with a null location? If so, just add a filter before the groupBy.

File: optaplanner-examples/src/test/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+        MrProcess process3 = new MrProcess();
+
+        MrProcessAssignment processAssignment1 = new MrProcessAssignment(0L, process1, machine1, machine2);
+        MrProcessAssignment processAssignment2 = new MrProcessAssignment(0L, process2, machine1, machine2);
+        MrProcessAssignment processAssignment3 = new MrProcessAssignment(0L, process3, machine2, machine1);
+
+        /*
+         * 2 processes are moving from machine1 to machine2, which has a cost of 20 => 2 * 20 * 10 (global penalty) = 400.
+         * The process3 moves from machine2 to machine1, which has a zero cost.
+         */
```
Comment (ge0ffrey) [inline optaplanner-examples/src/test/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProviderTest.java:299]: code style suggestion: this takes up 4 lines for 2 lines of content, typically we just use two `//` for 2 lines of non-javadoc comments.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java (hunk 6 above)
Comment (rsynek) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java:91]: See the discussion on Zulip chat - this is about uninitialized solution and the fact that a DRL score function works differently in such a condition.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 8]
```diff
@@ -145,74 +139,66 @@ private Constraint transientUsage(ConstraintFactory factory) {
...
-
-    private long balanceCost(MrBalancePenalty penalty, MrMachine machine, long originalUsage, long targetUsage) {
-        long originalAvailability =
-                machine.getMachineCapacity(penalty.getOriginResource()).getMaximumCapacity() - originalUsage;
-        long targetAvailability =
-                machine.getMachineCapacity(penalty.getTargetResource()).getMaximumCapacity() - targetUsage;
-        long balanceCost = penalty.getMultiplicand() * originalAvailability - targetAvailability;
-        return Math.max(0, balanceCost);
+    protected Constraint balanceCost(ConstraintFactory factory) {
+        throw new UnsupportedOperationException("Not yet implemented.");
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java:161]: Is removing this constraint intentional?

Comment (ge0ffrey) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java:161]: The old impl is wrong, so better to remove it.

**Pull request merged**